### ISSUE TITLE
Implement support for Diaspora nodes

### ISSUE THREAD

**Issue by bnvk** [snarfed/bridgy#198]
Diaspora has a strong community of users on various "pods" around the world http://podupti.me

While in Berlin, I've taken to manually POSSE'ing posts that might be of interest to hackers on one of their pods https://despora.de 

https://despora.de/posts/1951612
https://brennannovak.com/notes/472

Would be nice to backfeed posts from Disapora nodes!

**Comment by snarfed** [comment]
cc @phyks, who's expressed interest.

https://wiki.diasporafoundation.org/Federation_protocol_overview#ActivityStream_of_public_posts says you can get posts (and other data) as ActivityStreams, and that the reference implementation includes an API. yay.

**Comment by Phyks** [comment]
@snarfed 

Ok, I got some updates about Diaspora I think. First of all, there is a public feed per user, to publish its public posts. Mine is accessible at https://pod.phyks.me/public/phyks.atom. But this does not seem to be of great interest for the discussion here…

Next is [this page](https://wiki.diasporafoundation.org/Federation_protocol_overview) that you already linked. It says:
   
     An ActivityStream of public posts.
     An API that allows third-party applications to use your pod’s data on behalf of your users, using OAuth authentication (the OAuth protocol).

The ActivityStream of public post is the previous ATOM feed which does not include the necessary informations. However, I cannot find any information about the API. It is also mentionned at the end of [this old wiki page](https://github.com/diaspora/old_diaspora_wiki/blob/master/Federation-Protocol-Overview.md) but there is not much information.

Better, there is a feed associated with each posts. This means that for a post (say [this one](https://pod.phyks.me/posts/3834)), there is [an associated feed](https://pod.phyks.me/posts/3834.json) which includes `likes_count` and `reshares_count`. However, I do not know exactly how they are handled as on the previous example, I have a like, but it does not appear in `likes`. I fear that `likes` are internal to the pod only.

Finally, I found https://bitbucket.org/faldrian/diaspora-reshare-visualization/src which is the code behind http://jenseitsderfenster.de/diasporavisual/. If you enter a post (say https://pod.phyks.me/posts/3834), it gives you the number of likes, comments and reshares associated to this post, but not the actual information about them =(

**Comment by snarfed** [comment]
awesome, thanks for the research @Phyks!

btw, pubsub definitely isn't a requirement for bridgy, whether PuSH or anything else. we currently poll the four silos we do support, so I'd be fine with polling dispute sources too.

**Comment by snarfed** [comment]
showing only reply/like/reshare counts, and not the actual contents, is how the ActivityStreams schema works, so i'm not too surprised about that. it definitely seems like there should be some way to get the actual contents though. maybe at another URL endpoint?

for example, Google+'s API is built on ActivityStreams, and they have separate API calls to get replies, +1s, and reshares:
https://developers.google.com/+/api/latest/comments/list
https://developers.google.com/+/api/latest/people/listByActivity

**Comment by Phyks** [comment]
@snarfed 
Yep, I know that PuSH is not mandatory, but that would have been really nice :)

Actually, I have some doubts about Diaspora at the moment… I mean, it is an opensource project, but it seems to be even more closed than Facebook !

I idled around on their IRC channel, read the doc and all of this, in order to find more infos. But most of the doc is actually either out of date, purely descriptive, or talking about objectives in the far future (things like the OStatus API or so).

At the moment, I only see two ways:
* Implementing a scrapper.
* Using ActivityStreams.

But I cannot find any further doc about the provided ActivityStreams and about a particular other URL endpoint to fetch actual likes, replies and reshares.

So far, the only info I managed to get from them is ”read the f*cking source / doc” and “[here](https://github.com/Raven24/diaspora-federation) is the Ruby gem to handle federation, use it and that's all”. So I am quite disappointed by the community support at the moment, and feels like they are not interested at all to help us get backfeed =(

**Comment by snarfed** [comment]
that does sound disappointing. :(

I wonder if part of it is just that the project hasn't been very active for a while, since iirc the founders/leaders left a while back. that could explain everything being out of date and people not helping much.

**Comment by Phyks** [comment]
Yep, it is in some sort…

I really feel like the community is not interested at all in this possibility, and does not give a damn to crossposting to silos. Because Bridgy support could also give Diaspora the ability to handle crossposting to Facebook…

However, there are regularly commits to Diaspora repo, but mainly for backend stuff and nothing really huge and new, at least for the last six months.

At the moment, I feel like it will be heavily time consuming to manage to get a clean Diaspora support in Bridgy, and am not really convinced it is something really interesting, for any of the projects involved. At least, unless someone from Diaspora is willing to give some better infos…

Actually, I feel like Bridgy users have many expectations and this one may not be top priority, especially as I am not sure it concerns a bunch of people (but I may be underestimating).

I was interested in [Thingiverse](http://thingiverse.com/) backfeed as well, which offers a fully-featured API. I will focus on this first, to get first contact with Bridgy, I think :)

**Comment by tantek** [comment]
Phyks, I support Activity Streams 1.0 / Atom + PuSH on http://tantek.com/ with full discovery (basically the previous attempt at Federated Social Web standards) but don't know of any ability for Diaspora to consume this. For reference, Status.net *did* support this, and I was able to federate from my site to Status.net sites.

If you can find / demonstrate working (or interesting in) a Diaspora account "following" an arbitrary realtime AS1.0/Atom+PuSH feed, then there maybe something worth pursuing - I'd say that's the minimum bar that needs to be met by Diaspora before bothering to work on any Diaspora-specific support.

**Comment by singpolyma** [comment]
> Status.net did support this

And the continuation, GNU Social, still does.  I use @tantek's feed as well as my own to test such implementations all the time :)

**Comment by snarfed** [comment]
if anyone's interested in tackling this, i've written up [comprehensive instructions on how to add a new silo to bridgy.](https://github.com/snarfed/bridgy#adding-a-new-silo) feel free to jump in!

**Comment by Flaburgan** [comment]
Hi everyone, I'm from the diaspora* community and want to point that many things changed. First of all the federation has completely been rewritten and is not sitting at https://github.com/diaspora/diaspora_federation with the complete documentation here: https://diaspora.github.io/diaspora_federation/

We also hope to be able to provide an external API. The specifications are already written [here](https://diaspora.github.io/api-documentation/index.html) but are not implemented at the moment.

**Comment by snarfed** [comment]
fwiw, i'm now likely to work on diaspora support in https://fed.brid.gy/ first (which is close!), and less likely to do it here in bridgy.

**Comment by bekoli** [comment]
API has landed in `develop` [Announcement here](https://pod.diaspora.software/posts/707634201f9501387f3d5254001bd39e)

**Comment by snarfed** [comment]
Diaspora seems to be dying. https://diaspora.fediverse.observer/stats shows ~4k MAU (down over 99% from a year ago), ~50 servers.

cc snarfed/bridgy-fed#7
